Enable users to update approved claimed facility profiles & display claim info in the details sidebar

kellyi opened the pull request:
## Overview

This PR enables users with approved facility claims to edit these claims using a form on the /claimed/claimID route and also updates the facility details sidebar to display that info beneath the existing facility data.

I spun off the work of sending emails to a separate card.

I also spun off adding some multiselect tag fields to a separate card, since we don't really yet know the set of tags for these fields. It occurred to me that we can also remove & defer adding the `minimum_order_quantity` and `average_lead_time` fields, too, for now since users can currently just include this info in a free form way in the `facility_description` field.

Connects #518 

## Demo

![Screen Shot 2019-06-12 at 1 59 23 PM](https://user-images.githubusercontent.com/4165523/59374689-6e420f80-8d1a-11e9-9937-80a7a5f49306.png)


![Screen Shot 2019-06-11 at 9 13 22 PM](https://user-images.githubusercontent.com/4165523/59316728-e44c6500-8c8d-11e9-83fc-088df7634660.png)

## Testing Instructions

- get this branch, then `resetdb`, `processfixtures`, switch on the claim a facility and create a superuser
- sign up for an account through the UI, then fill out a claim a facility form
- sign in as the superuser and approve the claim
- sign back in as the other user and visit the OAR ID for the facility you have claimed. Clicking on the claim badge will now take you to your approved claim (if signed out, clicking on the badge will go to a not-yet-existing `/about/claimedfacilities` route
- verify that the claimed facility profile form loads and that you can edit and save the fields and that those changes get reflected accurately in the facility details UI

## Checklist

- [ ] `fixup!` commits have been squashed
- [ ] CI passes after rebase
- [ ] CHANGELOG.md updated with summary of features or fixes, following [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines

kellyi commented:
I am spinning off the multiselect fields for "certficiations", "product types", and "production types" to https://github.com/open-apparel-registry/open-apparel-registry/issues/576 since we may need to spend some time cataloging the set of options before adding those.

kellyi commented:
Remaining work here:

- [ ] email list holders when the claim a facility form has saved with new data
- [ ] link details verified badge to /about/claimedfacilities for users who have not claimed the facility and to /claimed/ID for users who have claimed the facility
- [ ] update the user serializer to send back a list of the OAR IDs for a user's pending & approved facility claims
- [ ] change country inputs to react select boxes

kellyi commented:
I spun off emailing contributors on profile changes to https://github.com/open-apparel-registry/open-apparel-registry/issues/578

jwalgran commented on src/app/src/components/FacilityDetailSidebar.jsx:
```diff
@@ -194,24 +217,53 @@ class FacilityDetailSidebar extends Component {
+                        </FeatureFlag>
                         <div className="control-panel__group">
                             <div style={detailsSidebarStyles.linkSectionStyle}>
                                 <a
```
The wireframes show conditionally hiding these links if the logged in contributor has claimed the facility, since the do not apply to the claimant. Instead, the wireframes have a link to the profile editing page above the satellite image. Since we already have access to `facilityIsClaimedByCurrentUser` we should consider implementing this link swapping.

kellyi replied:
Handled in be9749dfe4cbb9fbf4a31df0a72425441763cee5

Not my claimed facility:

<img width="463" alt="Screen Shot 2019-06-13 at 11 34 49 AM" src="https://user-images.githubusercontent.com/4165523/59446445-8162e700-8dcf-11e9-90bc-43e08d3da29b.png">

My claimed facility:

![Screen Shot 2019-06-13 at 11 34 54 AM](https://user-images.githubusercontent.com/4165523/59446459-8889f500-8dcf-11e9-9607-a664ad70459c.png)

kellyi replied:
Note that the verified badge will already also conditionally link to the facility claim form if it is my claimed facility.

jwalgran commented on src/app/src/components/ClaimedFacilitiesDetails.jsx:
```diff
@@ -1,9 +1,543 @@
+                    disabled={updating}
+                />
+                <InputSection
+                    label="Email"
```
Consider creating a follow up issue to add email address validation to this field.

kellyi replied:
This should be trivial to incorporate here, so I can just add it in this PR. I think we want to validate `isEmpty(email) || isEmail(email)` for this field, and probably disable the save button concurrently.

kellyi replied:
Added in 9c3296bf669dbadd3cc657cea27db06b0872f3ff

jwalgran commented on src/app/src/components/ClaimedFacilitiesDetails.jsx:
```diff
@@ -1,9 +1,543 @@
+                    onSwitchChange={updateFacilityPhoneVisibility}
+                />
+                <InputSection
+                    label="Website"
```
Consider creating a follow up issue to add url validation to this field and to render it as an active link in the sidebar (using the same conditional "http://" appending that we use on the contributor profile page.

kellyi replied:
This is/will be https://github.com/open-apparel-registry/open-apparel-registry/issues/587

jwalgran commented on src/app/src/components/ClaimedFacilitiesDetails.jsx:
```diff
@@ -1,9 +1,543 @@
+                    onChange={updateOfficePhone}
+                    disabled={updating}
+                />
+                {errorUpdating && (
```
Consider opening an issue to use `full_clean` on the model to trigger Django's model validation and generate helpful error messages that we can show to the claimant. We do this when parsing uploaded files https://github.com/open-apparel-registry/open-apparel-registry/blob/8fdc09c4d6cac5f4aefc3a313e9c04d84e5ab49f/src/django/api/processing.py#L127

We are currently returning 500 when the claimant enters too many characters into a field.

<img width="1680" alt="Screen Shot 2019-06-12 at 8 18 06 PM" src="https://user-images.githubusercontent.com/17363/59401488-b62f5980-8d4f-11e9-8a66-f40be34deae3.png">

We could also make use of `maxlength` to prevent these types of errors.

kellyi replied:
I made https://github.com/open-apparel-registry/open-apparel-registry/issues/589 for handling it. I think we could also use the existing serializer to check somehow.

jwalgran commented on src/app/src/components/ClaimedFacilitiesDetails.jsx:
```diff
@@ -1,9 +1,543 @@
+                        color="primary"
+                        disabled={updating}
+                    >
+                        Submit
```
Consider "Save" or "Save Changes" instead of "Submit"

Also consider showing some "toast" or similar if there is a 200 response. Right now there is no indication that the save succeeded.

kellyi replied:
"Save" is in 86904d276ce2250c4697174658b0f20ef45dc6ae

toast is in cd8cc08bf7b3a8a116313adf68a17112df28c48c

kellyi commented on src/app/src/components/FacilityDetailSidebar.jsx:
```diff
@@ -146,20 +139,38 @@ class FacilityDetailSidebar extends Component {
                         </h3>
                         <p className="panel-header__subheading notranslate">
-                            {data.properties.address} - {data.properties.country_name}
+                            {data.properties.address} -{' '}
```
I think this got added by running Prettier on this file in my editor; it's how Prettier preserves displayed space chars when breaking text across lines. I can put this all on a single line, though, and remove the `{' '}`.

kellyi replied:
Done in 1bf02ca7e5e4ddecf400d97da3ea2b79a5a7f7e8

kellyi commented on src/app/src/components/FacilityDetailSidebarClaimedInfo.jsx:
```diff
@@ -0,0 +1,120 @@
+
+            <ClaimInfoSection
+                label="Claimed Facility Address"
+                value={`${facility.address || ' '} ${facility.country || ' '}`}
```
Good point. For now I'll remove the facility country selection and field altogether from the form & model (but leave the one for office country). If we decide we do want to allow changing a facility country, then we can add it back later.

kellyi replied:
Removed in 9c3296bf669dbadd3cc657cea27db06b0872f3ff

kellyi commented on src/app/src/components/FacilityDetailSidebarClaimedInfo.jsx:
```diff
@@ -0,0 +1,120 @@
+                label="Claimed Facility Phone Number"
+            />
+            <ClaimInfoSection
+                value={facility.minimum_order}
```
Fixed in 9c3296bf669dbadd3cc657cea27db06b0872f3ff

kellyi commented on src/app/src/components/FacilityDetailSidebarClaimedInfo.jsx:
```diff
@@ -0,0 +1,120 @@
+            />
+
+            <ClaimInfoSection
+                label="Claimed Facility Description"
```
Done in 3c0b2191be162f5df1a65cdf63a1007603e06bbc

I also added a header to indicate what the info represents. I think we will also need to add some kind of disclaimed or link to the about page here.

kellyi replied:
![Screen Shot 2019-06-13 at 11 04 45 AM](https://user-images.githubusercontent.com/4165523/59444068-36df6b80-8dcb-11e9-9e84-cb803d5df8f7.png)

kellyi replied:
Disclaimer message issue is https://github.com/open-apparel-registry/open-apparel-registry/issues/588

jwalgran replied:
Consider dropping the "Facility" prefix too. All the information in the sidebar is about the facility.

kellyi replied:
Dropped the "Facility" prefix in e374b2d275b6f9d901fe9bb3628a3de475015102

![Screen Shot 2019-06-13 at 11 21 37 AM](https://user-images.githubusercontent.com/4165523/59445368-89218c00-8dcd-11e9-9a6d-be4404fba88a.png)

kellyi commented on src/django/api/serializers.py:
```diff
@@ -99,6 +101,24 @@ def get_contributor_id(self, user):
+            return []
+
+        try:
+            contributor = Contributor.objects.get(admin=user)
```
Thanks for the suggestion! This is in 5d2e4dead336f82ec85898f7836ce331d3f615ad

kellyi commented on src/django/api/views.py:
```diff
@@ -1313,8 +1313,10 @@ def list(self, request):
 
-        response_data = FacilityClaimSerializer(FacilityClaim.objects.all(),
-                                                many=True).data
+        response_data = FacilityClaimSerializer(
```
I'll add this before merging.

kellyi commented:
@jwalgran I believe I've addressed all the feedback above, so this is ready for another look!

I updated the new migration, so testing will need to start from `resetdb` again.

kellyi commented:
I'm going to squash this down again.

jwalgran commented:
Thanks for all the revisions. Taking another look.

jwalgran commented:
Running into a prop type error when selecting a facility on the map

![2019-06-13 09 05 06](https://user-images.githubusercontent.com/17363/59448630-6c2f8d80-8dba-11e9-82de-2196fbcb43c2.gif)

```
Warning: Failed prop type: The prop `data` is marked as required in `FacilityDetailsSidebarClaimedInfo`, but its value is `null`.
    in FacilityDetailsSidebarClaimedInfo (at FacilityDetailSidebar.jsx:221)
    in FacilityDetailSidebar (created by Connect(FacilityDetailSidebar))
    ...
```

kellyi commented:
Fixed the propType error in d53d5c43eea285853f90c20fcdc40afd4f695cb4

kellyi commented:
Awesome, thanks for all your help with this!
